src/helpers/buffers/unordered_receiver.rs

```diff
@@ -203,30 +215,78 @@ where
 
     /// Poll for the next record.  This should only be invoked when
     /// the future for the next message is polled.
+    ///
+    /// ## Errors
+    /// If buffer capacity is not enough to read `M` and the underlying stream does not have
+    /// more data. This may lead to different behavior depending on the order of issued reads. See
+    /// [`read_order`] test for an example.
+    ///
+    /// [`read_order`]: test::read_order
     fn poll_next<M: Message>(&mut self, cx: &mut Context<'_>) -> Poll<Result<M, Error>> {
-        if let Some(m) = self.spare.read() {
+        // If spare has enough data for us, poll it first
+        // otherwise, poll the underlying stream until it returns pending or it provides enough
+        // data to return a value.
+        let message = self.spare.read();
+        let next = if let Some(m) = message {
+            // this check exists to make sure the inner stream is eventually moved to
+            // the closed state. We don't want to poll it too often, but we also need to know
+            // when it is done and `UnorderedReceiver` can be dropped.
             self.wake_next();
-            return Poll::Ready(Ok(m));
-        }
-
-        loop {
-            match self.stream.as_mut().poll_next(cx) {
-                Poll::Pending => {
-                    return Poll::Pending;
-                }
-                Poll::Ready(Some(b)) => {
-                    if let Some(m) = self.spare.extend(b.as_ref()) {
+            Ok(m)
+        } else {
+            loop {
+                if let Some(bytes) = ready!(self.stream.as_mut().poll_next(cx)) {
+                    if let Some(m) = self.spare.extend(bytes.as_ref()) {
                         self.wake_next();
-                        return Poll::Ready(Ok(m));
+                        break Ok(m);
                     }
+                } else {
+                    self.spare.replace(&[]);
+                    self.close();
+                    break Err(Error::EndOfStream {
+                        record_id: RecordId::from(self.next),
+                    });
+                }
+            }
+        };
+
+        if next.is_ok() && self.spare.is_empty() {
+            // we don't want to be woken up here, control loop is driven by the client.
+            // They decide when they want the next message and must issue a `poll` for it.
```

Comment:
this is a bug - at the end of the stream if stream is not ready, nothing will be polled